pennylane_lightning/experimental/lightning_qubit_2.py

```diff
@@ -0,0 +1,171 @@
+
+        def preprocess(
+            self,
+            circuits: QuantumTape_or_Batch,
+            execution_config: ExecutionConfig = DeviceExecutionConfig,
+        ) -> Tuple[QuantumTapeBatch, Callable]:
+            """Converts an arbitrary circuit or batch of circuits into a batch natively executable by the :meth:`~.execute` method.
+
+            Args:
+                circuits (Union[QuantumTape, Sequence[QuantumTape]]): The circuit or a batch of circuits to preprocess
+                    before execution on the device
+                execution_config (Union[ExecutionConfig, Sequence[ExecutionConfig]]): A data structure describing the parameters needed to fully describe
+                    the execution. Includes such information as shots.
+
+            Returns:
+                Sequence[QuantumTape], Callable: QuantumTapes that the device can natively execute
+                and a postprocessing function to be called after execution.
+
+            This device:
+
+            * Supports any qubit operations that provide a matrix
+            * Currently does not support finite shots
+            * Currently does not intrinsically support parameter broadcasting
+
+            """
+            is_single_circuit = False
+            if isinstance(circuits, QuantumScript):
+                circuits = [circuits]
+                is_single_circuit = True
+
+            batch, post_processing_fn = preprocess(circuits, execution_config=execution_config)
+
+            if is_single_circuit:
+
+                def convert_batch_to_single_output(results):
+                    """Unwraps a dimension so that executing the batch of circuits looks like executing a single circuit."""
+                    return post_processing_fn(results)[0]
+
+                return batch, convert_batch_to_single_output
+
+            return batch, post_processing_fn
+
+        def execute(
+            self,
+            circuits: QuantumTape_or_Batch,
+            execution_config: ExecutionConfig = DeviceExecutionConfig,
+        ):
+            """Execute a circuit or a batch of circuits and turn it into results.
+
+            Args:
+                circuits (Union[QuantumTape, Sequence[QuantumTape]]): the quantum circuits to be executed
+                execution_config (Union[ExecutionConfig, Sequence[ExecutionConfig]]): a data structure with additional information required for execution
+
+            Returns:
+                TensorLike, tuple[TensorLike], tuple[tuple[TensorLike]]: A numeric result of the computation.
+            """
+            is_single_circuit = False
+            if isinstance(circuits, QuantumScript):
+                is_single_circuit = True
+                circuits = [circuits]
+
+            if self.tracker.active:
+                self.tracker.update(batches=1, executions=len(circuits))
+                self.tracker.record()
+
+            results = tuple(simulate(c) for c in circuits)
```

Comment:
The nice thing about here is that we can choose to use Ray, Dask, `concurrent.futures` compatible context managers and the like as a follow-up extension.